File: apps/cargo-tracker/app/cargotracker/shared/domain/Money.scala
package cargotracker.shared.domain

/** 多通貨対応の金額。
  *
  *   - `amount` は **最小通貨単位**（JPY=円、USD=cent）の整数で保持する
  *   - 等価性は `case class` の構造比較を利用する
  */
final case class Money private (currency: String, amount: Long):

  /** 同一通貨同士で加算する。 */
  def +(other: Money): Either[Money.Error, Money] =
    if currency != other.currency then Left(Money.CurrencyMismatch)
    else Right(new Money(currency, amount + other.amount))

  /** 数量倍する。 */
  def times(factor: Long): Money = copy(amount = amount * factor)

object Money:

  sealed trait Error
  case object NegativeAmount extends Error
  case object CurrencyMismatch extends Error

  /** 任意通貨の Money を生成する。 */
  def apply(currency: String, amount: Long): Either[Error, Money] =
    if amount < 0 then Left(NegativeAmount)
    else Right(new Money(currency, amount))

  /** 日本円の便利コンストラクタ。 */
  def jpy(amount: Long): Either[Error, Money] = apply("JPY", amount)

  val zeroJpy: Money = new Money("JPY", 0L)

File: docs/article/source/haskell/apps/cargo-tracker/src/Cargotracker/Billing/Domain/Model/Value/Money.hs
{- | 金額 VO (US23, IT8, Billing BC)

data-model.md 設計判断 3 に従い、最小通貨単位の Integer で保持する
(`Double` は金額計算に使用しない)。通貨は ISO 4217 の 3 文字大文字コード。

Pricing BC の `Cost` と同型だが、Rule 4 (BC 間 Domain 直接参照禁止) に
従い Billing BC 専用の VO として定義する。
-}
module Cargotracker.Billing.Domain.Model.Value.Money
  ( Money (..),
    mkMoney,
    zeroMoney,
  ) where

import Data.Char (isAsciiUpper)
import Data.Text (Text)
import qualified Data.Text as T

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data Money = Money
  { moneyAmount :: !Integer
  -- ^ 最小通貨単位 (JPY なら円、USD ならセント)
  , moneyCurrency :: !Text
  -- ^ ISO 4217 3 文字大文字コード
  }
  deriving stock (Eq, Show)

mkMoney :: Integer -> Text -> Either DomainError Money
mkMoney amount currency
  | amount < 0 = Left (InvalidCost amount)
  | not validCurrency = Left (InvalidCurrency currency)
  | otherwise = Right (Money amount currency)
  where
    validCurrency = T.length currency == 3 && T.all isAsciiUpper currency

zeroMoney :: Text -> Either DomainError Money
zeroMoney = mkMoney 0
